massa-pool-worker/src/denunciation_pool.rs

```diff
@@ -1,31 +1,51 @@
-use massa_models::denunciation::{Denunciation, DenunciationIndex};
-use massa_models::prehash::PreHashMap;
+use std::collections::{hash_map::Entry, HashMap};
+use tracing::{debug, info, warn};
 
+use massa_models::slot::Slot;
+use massa_models::{
+    address::Address,
+    denunciation::{
+        BlockHeaderDenunciationPrecursor, Denunciation, DenunciationPrecursor,
+        EndorsementDenunciationPrecursor,
+    },
+    timeslots::get_closest_slot_to_timestamp,
+};
 use massa_pool_exports::PoolConfig;
+use massa_pos_exports::SelectorController;
+use massa_time::MassaTime;
 
 pub struct DenunciationPool {
     /// configuration
     config: PoolConfig,
-    /// storage
-    // storage: Storage,
+    /// selector controller to get draws
+    pub selector: Box<dyn SelectorController>,
     /// last consensus final periods, per thread
     last_cs_final_periods: Vec<u64>,
-    /// internal cache
-    denunciations_cache: PreHashMap<DenunciationIndex, Denunciation>,
+    // /// internal cache
+    // denunciations_cache: PreHashMap<DenunciationIndex, Denunciation>,
+    /// Internal cache for endorsement denunciation
+    /// store at most 1 endorsement per entry, as soon as we have 2 we produce a Denunciation
+    endorsements_by_slot_index: HashMap<(Slot, u32), EndorsementDenunciationStatus>,
+    /// Internal cache for block header denunciation
+    block_header_by_slot: HashMap<Slot, BlockHeaderDenunciationStatus>,
 }
 
```

Comment:
I think the two hashmaps can be unified into one by defining:

```
#[derive(Hash)]
enum DenunciationTarget {
    Block {
        slot: Slot,
    },
    Endorsement: {
        slot: Slot,
        index: u64
    }
}
```

and change `DenunciationStatus` to:

```
enum DenunciationStatus {
    Accumulating(DenunciationPrecursor),
    DenunciationEmitted(Denunciation),
}
```

then we could use:

```
denunciations: HashMap<DenunciationTarget, DenunciationStatus>
```